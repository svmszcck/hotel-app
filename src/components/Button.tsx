import React from "react";
import { Button, GestureResponderEvent } from "react-native";

type ButtonVariant = "primary" | "secondary" | "danger";

type ButtonProps = {
  title: string;
  onPress: (event: GestureResponderEvent) => void;
  variant?: ButtonVariant;
};

const CustomButton = ({ onPress, title, variant = "primary" }: ButtonProps) => {
  return (
    <Button
      onPress={onPress}
      title={title}
      color="#841584"
      accessibilityLabel="Learn more about this purple button"
    />
  );
};

export default CustomButton;
