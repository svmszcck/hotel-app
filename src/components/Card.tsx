import React, { FC, ReactNode } from "react";
import { Text, View, StyleSheet, ViewStyle } from "react-native";

import Colors from "constants/colors";
import Spacing from "constants/spacing";

type CardProps = {
  text: string;
  style?: ViewStyle;
};

const Card: FC<CardProps> = ({ text, style }) => {
  return (
    <View style={StyleSheet.flatten([styles.container, style])}>
      <Text>{text}</Text>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    alignSelf: "flex-start",
    padding: Spacing.MEDIUM,
    backgroundColor: Colors.GREEN,
  },
});

export default Card;
