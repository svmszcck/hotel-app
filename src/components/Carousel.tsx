import React, { useMemo, FC } from "react";
import { Image, View, StyleSheet } from "react-native";
import { SwiperFlatList } from "react-native-swiper-flatlist";
import { SharedElement } from "react-navigation-shared-element";

import { screenWidth } from "utils/ui";
import { trimImages } from "utils/general";
import { PropertyImg } from "global-types";

import placeholder from "assets/images/placeholder.png";

const width = screenWidth();

type CarouselProps = {
  images: PropertyImg[];
};

const Carousel: FC<CarouselProps> = ({ images }) => {
  const trimmedImages = useMemo(() => trimImages(images), [images]);

  return (
    <View style={styles.container}>
      <SwiperFlatList
        showPagination
        data={trimmedImages}
        keyExtractor={(item) => item.url}
        renderItem={({ item }) => {
          return (
            <View style={styles.child} testID="carousel-item">
              <SharedElement id={`hotel-photo-${item.position}`}>
                <Image
                  source={item.url ? { uri: item.url } : placeholder}
                  style={styles.cover}
                  testID="carousel-item-image"
                />
              </SharedElement>
            </View>
          );
        }}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    backgroundColor: "white",
  },
  child: {
    width,
    justifyContent: "center",
  },
  cover: {
    width: "100%",
    height: 250,
    minHeight: 300,
  },
});

export default Carousel;
