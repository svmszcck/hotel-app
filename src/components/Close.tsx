import React, { FC, useState, useEffect } from "react";
import {
  Text,
  StyleSheet,
  GestureResponderEvent,
  TouchableOpacity,
} from "react-native";

import globalStyles from "global-styles";

type CloseProps = {
  onPress: (event: GestureResponderEvent) => void;
};

const Close: FC<CloseProps> = ({ onPress }) => {
  const [close, showClose] = useState<boolean>(false);

  useEffect(() => {
    setTimeout(() => showClose(true), 500);
  }, []);

  return close ? (
    <TouchableOpacity
      style={StyleSheet.flatten([globalStyles.close, { top: 15, left: 15 }])}
      onPress={onPress}
      activeOpacity={0.8}
    >
      <Text>X</Text>
    </TouchableOpacity>
  ) : null;
};

export default Close;
