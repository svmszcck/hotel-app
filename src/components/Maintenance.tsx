import React from "react";
import { Image, View, Text, StyleSheet } from "react-native";

import vacation from "assets/images/vacation.png";

const Maintenance = () => {
  return (
    <View style={styles.container}>
      <Image source={vacation} style={styles.vector} resizeMode="cover" />
      <Text style={styles.text}>Not Implemented Yet :(</Text>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
  },
  vector: {
    justifyContent: "center",
    alignItems: "center",
    width: 300,
    height: 200,
  },
  text: {
    fontWeight: "600",
    fontSize: 18,
    marginTop: 20,
  },
});

export default Maintenance;
