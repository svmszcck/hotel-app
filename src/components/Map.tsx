import React, { FC, ReactNode } from "react";
import { Text, View, StyleSheet } from "react-native";
import MapView, { Callout, Marker, PROVIDER_DEFAULT } from "react-native-maps";
import { useContext } from "use-context-selector";
import Icon from "@expo/vector-icons/FontAwesome";

import { getRandomInt } from "utils/general";
import { screenWidth, screenHeight } from "utils/ui";
import GlobalContext from "contexts/GlobalContext";
import Colors from "constants/colors";
import Spacing from "constants/spacing";
import {
  DEFAULT_LOCATION,
  MARKER_PRESS,
  MIN_PRICE,
  MAX_PRICE,
} from "constants/general";
import { Property } from "global-types";

type MapProps = {};

const renderMarker: FC = (marker: Property, isActive: boolean) => {
  return (
    <>
      <View style={styles.markerWrapper}>
        <View
          style={StyleSheet.flatten([
            styles.marker,
            {
              backgroundColor: isActive ? Colors.ORANGE : Colors.GRAY_DARK,
            },
          ])}
        >
          <Text style={styles.markerText}>
            {marker.lowest_price_per_night | getRandomInt(MIN_PRICE, MAX_PRICE)}
            â‚¬
          </Text>
        </View>
        <Icon
          name="caret-down"
          size={20}
          color={isActive ? Colors.ORANGE : Colors.GRAY_DARK}
          style={styles.arrowIcon}
        />
      </View>

      <Callout tooltip>
        <View></View>
      </Callout>
    </>
  );
};

const Map: FC<MapProps> = () => {
  const state = useContext(GlobalContext);
  const { properties, activeProperty, setActiveProperty } = state;

  const handleClick = (id: number) => {
    if (activeProperty && id === activeProperty) setActiveProperty(null);
    else setActiveProperty(id);
  };

  return (
    <View>
      <MapView
        style={styles.map}
        provider={PROVIDER_DEFAULT}
        zoomEnabled
        initialRegion={DEFAULT_LOCATION}
        onPress={(e) =>
          e.nativeEvent.action !== MARKER_PRESS ? setActiveProperty(null) : null
        }
      >
        {properties.map((marker, index) => {
          return (
            <Marker
              key={index}
              coordinate={{
                latitude: marker.location.lat,
                longitude: marker.location.lng,
              }}
              onPress={() => handleClick(marker.id)}
              zIndex={marker.id === activeProperty ? 1000 : 0}
            >
              {renderMarker(marker, marker.id === activeProperty)}
            </Marker>
          );
        })}
      </MapView>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#fff",
    alignItems: "center",
    justifyContent: "center",
  },
  map: {
    width: screenWidth(),
    height: screenHeight(),
  },
  markerWrapper: {
    flex: 1,
    alignItems: "center",
  },
  marker: {
    width: 40,
    height: 40,
    justifyContent: "center",
    alignItems: "center",
    borderRadius: Spacing.SMALL,
  },
  markerText: {
    color: "white",
  },
  arrowIcon: {
    marginTop: -8,
  },
});

export default Map;
