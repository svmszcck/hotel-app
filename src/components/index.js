export { default as Slider } from "./Slider";
export { default as Button } from "./Button";
export { default as Card } from "./Card";
export { default as Map } from "./Map";
export { default as HotelPanel } from "./HotelPanel";
export { default as Separator } from "./Separator";
export { default as ScreenContainer } from "./ScreenContainer";
export { default as Close } from "./Close";
export { default as SafeAreaView } from "./SafeAreaView";
export { default as Score } from "./Score";
export { default as Maintenance } from "./Maintenance";
export { default as Carousel } from "./Carousel";
