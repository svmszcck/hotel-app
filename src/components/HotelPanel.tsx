import React, { FC } from "react";
import { Image, View, Text, StyleSheet, Pressable } from "react-native";
import { useNavigation } from "@react-navigation/native";
import Ionicon from "@expo/vector-icons/Ionicons";
import { SharedElement } from "react-navigation-shared-element";

import Colors from "constants/colors";
import Spacing from "constants/spacing";
import Routes from "constants/routes";
import { Separator, Score } from "components";
import usePropertySelect from "hooks/usePropertySelect";

import placeholder from "assets/images/placeholder.png";

const HotelPanel: FC = () => {
  const navigation = useNavigation();
  const activeProperty = usePropertySelect();
  const coverImage = activeProperty?.images[0]?.url;

  return (
    <Pressable
      style={styles.container}
      testID="hotel-panel-container"
      onPress={() => navigation.navigate(Routes.DETAILS as never)}
    >
      <SharedElement id="hotel-photo-0">
        <View style={styles.imageWrapper}>
          <Image
            source={coverImage ? { uri: coverImage } : placeholder}
            style={styles.image}
            resizeMode="cover"
          />
        </View>
      </SharedElement>
      <Score style={styles.score} />
      <View style={styles.info}>
        <Text style={styles.title} numberOfLines={1}>
          {activeProperty?.name}
        </Text>
        <View style={styles.location}>
          <Ionicon
            name="location-outline"
            size={20}
            color={Colors.ORANGE}
            style={styles.locationIcon}
          />
          <Text numberOfLines={2}>
            {activeProperty?.distance.toFixed(1)} km from city center
          </Text>
        </View>
        <Separator style={styles.separator} />
        {activeProperty?.lowest_price_per_night ? (
          <Text>
            From{" "}
            <Text style={styles.price}>
              {activeProperty?.lowest_price_per_night}â‚¬/Night
            </Text>
          </Text>
        ) : (
          <Text>No Price Info</Text>
        )}
      </View>
    </Pressable>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    flex: 1,
  },
  imageWrapper: {
    backgroundColor: Colors.BLACK,
  },
  image: {
    width: 100,
    height: "100%",
    opacity: 0.6,
  },
  info: {
    flex: 1,
    padding: Spacing.MEDIUM,
  },
  location: {
    flexDirection: "row",
    alignItems: "center",
    marginTop: Spacing.MEDIUM,
  },
  locationIcon: {
    marginRight: Spacing.SMALL,
  },
  title: {
    fontSize: 18,
    fontWeight: "500",
  },
  price: {
    color: Colors.ORANGE,
    fontWeight: "600",
  },
  separator: {
    marginBottom: Spacing.MEDIUM,
  },
  score: {
    left: 30,
    right: undefined,
  },
});

export default HotelPanel;
