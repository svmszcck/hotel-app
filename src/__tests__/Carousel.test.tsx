import { render, screen, fireEvent } from "@testing-library/react-native";
import renderer from "react-test-renderer";

import { Carousel } from "components";

describe("Carousel Panel Component", () => {
  const images = [
    {
      url: "https://i.picsum.photos/id/270/536/354.jpg?hmac=mnhmFHibNPisjwBAA_oFyczU3QOuP0Pd6-oeEm5IQaU",
      is_portrait: true,
      position: 0,
      unit_group_ids: [],
      tags: [],
    },
  ];
  test("renders correctly", () => {
    render(<Carousel images={images} />);
    const rendered = renderer.create(<Carousel images={images} />).toJSON();

    expect(screen.getAllByTestId("carousel-item").length).toBe(1);
    expect(screen.getByTestId("carousel-item-image")).toBeTruthy();
    expect(rendered).toMatchSnapshot();
  });
});
