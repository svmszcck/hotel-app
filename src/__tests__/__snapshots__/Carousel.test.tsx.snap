// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel Panel Component renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "is_portrait": true,
          "position": 0,
          "tags": Array [],
          "unit_group_ids": Array [],
          "url": "https://i.picsum.photos/id/270/536/354.jpg?hmac=mnhmFHibNPisjwBAA_oFyczU3QOuP0Pd6-oeEm5IQaU",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    initialNumToRender={1}
    initialScrollIndex={0}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    onScrollToIndexFailed={[Function]}
    onViewableItemsChanged={[Function]}
    pagingEnabled={true}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEnabled={true}
    scrollEventThrottle={50}
    showsHorizontalScrollIndicator={false}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    viewabilityConfig={
      Object {
        "itemVisiblePercentThreshold": 60,
        "minimumViewTime": 200,
      }
    }
    viewabilityConfigCallbackPairs={
      Array [
        Object {
          "onViewableItemsChanged": [Function],
          "viewabilityConfig": Object {
            "itemVisiblePercentThreshold": 60,
            "minimumViewTime": 200,
          },
        },
      ]
    }
  >
    <View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          style={
            Object {
              "justifyContent": "center",
              "width": 750,
            }
          }
          testID="carousel-item"
        >
          <View
            collapsable={false}
          >
            <Image
              source={
                Object {
                  "uri": "https://i.picsum.photos/id/270/536/354.jpg?hmac=mnhmFHibNPisjwBAA_oFyczU3QOuP0Pd6-oeEm5IQaU",
                }
              }
              style={
                Object {
                  "height": 250,
                  "minHeight": 300,
                  "width": "100%",
                }
              }
              testID="carousel-item-image"
            />
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Array [
        Object {
          "alignSelf": "center",
          "bottom": 0,
          "flexDirection": "row",
          "height": 28.125,
          "justifyContent": "center",
          "marginVertical": 16.675,
          "position": "absolute",
        },
        Object {},
      ]
    }
  >
    <View
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 25,
          "height": 28.125,
          "marginHorizontal": 18.75,
          "opacity": 1,
          "width": 28.125,
        }
      }
      testID="_pagination_0"
    />
  </View>
</View>
`;
